File: 623-add-one-row-to-tree/623-add-one-row-to-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func addOneRow(root *TreeNode, val int, depth int) *TreeNode {
    if root == nil {
        return nil
    }
    if depth == 1 {
        return &TreeNode{ Val: val, Left: root }
    }
    if depth == 2 {
        root.Left = &TreeNode{ Val: val, Left: root.Left }
        root.Right = &TreeNode{ Val: val, Right: root.Right }
        return root
    }
    addOneRow(root.Left, val, depth - 1)
    addOneRow(root.Right, val, depth - 1)
    return root
}

File: 623-add-one-row-to-tree/623-add-one-row-to-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} val
 * @param {number} depth
 * @return {TreeNode}
 */
var addOneRow = function(root, val, depth) {
  if (!root) {
    return;
  }
  if (depth === 1) {
    return new TreeNode(val, root, null);
  }
  if (depth === 2) {
    root.left = new TreeNode(val, root.left, null);
    root.right = new TreeNode(val, null, root.right);
    return root;
  }
  root.left && addOneRow(root.left, val, depth - 1);
  root.right && addOneRow(root.right, val, depth - 1);
  return root;
};